### PR TITLE
Fix nested list indexing when the index is 0 or larger than the list size

### PR DESCRIPTION
**Pull request by ahmedriza**
# Which issue does this PR close?

Closes https://github.com/apache/arrow-datafusion/issues/5310

# Rationale for this change
Indexing a nested list should not fail even when the index is out of range

# What changes are included in this PR?
Correctly handle the case when an index is out or range, i.e. 0 or larger than the value of the list

# Are these changes tested?
Yes

# Are there any user-facing changes?
No

### CODE REVIEW

File: datafusion/physical-expr/src/expressions/get_indexed_field.rs [hunk 1]
```diff
@@ -390,4 +404,78 @@ mod tests {
...
+        arg: Arc<dyn PhysicalExpr>,
+        index: i64,
+        expected_result: ArrayRef,
+    ) {
+        let expr = Arc::new(GetIndexedFieldExpr::new(
+            arg,
+            ScalarValue::Int64(Some(index)),
+        ));
+        let result = expr.evaluate(batch).unwrap().into_array(batch.num_rows());
+        assert!(
```
Comment (alamb) [inline datafusion/physical-expr/src/expressions/get_indexed_field.rs:458]: the `assert_batches_eq!` macro is often used for such comparisons too

File: datafusion/physical-expr/src/expressions/get_indexed_field.rs [hunk 2]
```diff
@@ -390,4 +404,78 @@ mod tests {
...
+        let col_a = col("a", &schema).unwrap();
+        // out of bounds index
+        verify_index_evaluation(&batch, col_a.clone(), 0, float64_array(None));
+
+        verify_index_evaluation(&batch, col_a.clone(), 1, float64_array(Some(1.0)));
+        verify_index_evaluation(&batch, col_a.clone(), 2, float64_array(None));
+        verify_index_evaluation(&batch, col_a.clone(), 3, float64_array(Some(3.0)));
+
+        // out of bounds index
+        verify_index_evaluation(&batch, col_a.clone(), 100, float64_array(None));
```
Comment (alamb) [inline datafusion/physical-expr/src/expressions/get_indexed_field.rs:444]: 👍

File: datafusion/physical-expr/src/expressions/get_indexed_field.rs [hunk 3]
```diff
@@ -390,4 +404,78 @@ mod tests {
...
+                let mut b = PrimitiveBuilder::<Float64Type>::new();
+                b.append_value(v);
+                Arc::new(b.finish())
+            }
+            None => {
+                let mut b = PrimitiveBuilder::<Float64Type>::new();
+                b.append_null();
+                Arc::new(b.finish())
+            }
+        }
```
Comment (alamb) [inline datafusion/physical-expr/src/expressions/get_indexed_field.rs:479]: I think you can make a Float64Array directly from the value if you wanted:


```rust
Arc::new(Float64Array::from_value(value, 1))
```

https://docs.rs/arrow/33.0.0/arrow/array/struct.PrimitiveArray.html#method.from_value

**Review by alamb — APPROVED** [review]
LGTM -- thank you @ahmedriza 


I ran the new test without the changes in this PR and it fails as expected. 👍 

```

---- expressions::get_indexed_field::tests::get_indexed_field_list_out_of_bounds stdout ----
thread 'expressions::get_indexed_field::tests::get_indexed_field_list_out_of_bounds' panicked at 'result: ListArray
[
  null,
] != expected result: PrimitiveArray<Float64>
[
  null,
]', datafusion/physical-expr/src/expressions/get_indexed_field.rs:451:9
```

File: datafusion/physical-expr/src/expressions/get_indexed_field.rs (hunk 1 above)
Comment (ahmedriza) [reply datafusion/physical-expr/src/expressions/get_indexed_field.rs:458]: Thanks @alamb.  Didn't know about this macro. Can see it in `datafusion::core::test_util.rs`.  However, can't see `core` being a dependency of `datafusion-physical-expr` where this test is.  Perhaps I've misunderstood what was meant.

**Comment by alamb** [comment]
Thanks again @ahmedriza

**Pull request merged**